testutils/sims/runner.go

```diff
@@ -0,0 +1,233 @@
+package sims
+
+import (
+	"fmt"
+	"io"
+	"path/filepath"
+	"testing"
+
+	dbm "github.com/cosmos/cosmos-db"
+	"github.com/stretchr/testify/require"
+
+	"cosmossdk.io/log"
+
+	"github.com/cosmos/cosmos-sdk/baseapp"
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/client/flags"
+	"github.com/cosmos/cosmos-sdk/codec"
+	"github.com/cosmos/cosmos-sdk/runtime"
+	"github.com/cosmos/cosmos-sdk/server"
+	servertypes "github.com/cosmos/cosmos-sdk/server/types"
+	simtestutil "github.com/cosmos/cosmos-sdk/testutil/sims"
+	simtypes "github.com/cosmos/cosmos-sdk/types/simulation"
+	"github.com/cosmos/cosmos-sdk/x/simulation"
+	"github.com/cosmos/cosmos-sdk/x/simulation/client/cli"
+)
+
+const SimAppChainID = "simulation-app"
+
+// this list of seeds was imported from the original simulation runner: https://github.com/cosmos/tools/blob/v1.0.0/cmd/runsim/main.go#L32
+var defaultSeeds = []int64{
+	1, 2, 4, 7,
+	32, 123, 124, 582, 1893, 2989,
+	3012, 4728, 37827, 981928, 87821, 891823782,
+	989182, 89182391, 11, 22, 44, 77, 99, 2020,
+	3232, 123123, 124124, 582582, 18931893,
+	29892989, 30123012, 47284728, 7601778, 8090485,
+	977367484, 491163361, 424254581, 673398983,
+}
+
+type SimStateFactory struct {
+	Codec       codec.Codec
+	AppStateFn  simtypes.AppStateFn
+	BlockedAddr map[string]bool
+}
+
+// SimulationApp abstract app that is used by sims
+type SimulationApp interface {
+	runtime.AppI
+	SetNotSigverifyTx()
+	GetBaseApp() *baseapp.BaseApp
+	TxConfig() client.TxConfig
+}
+
+// Run is a helper function that runs a simulation test with the given parameters.
+// It calls the RunWithSeeds function with the default seeds and parameters.
+//
+// This is the entrypoint to run simulation tests that used to run with the runsim binary.
+func Run[T SimulationApp](
+	t *testing.T,
+	appFactory func(
+		logger log.Logger,
+		db dbm.DB,
+		traceStore io.Writer,
+		loadLatest bool,
+		appOpts servertypes.AppOptions,
+		baseAppOptions ...func(*baseapp.BaseApp),
+	) T,
+	setupStateFactory func(app T) SimStateFactory,
+	postRunActions ...func(t *testing.T, app TestInstance[T]),
+) {
+	t.Helper()
+	RunWithSeeds(t, appFactory, setupStateFactory, defaultSeeds, nil, postRunActions...)
+}
+
+// RunWithSeeds is a helper function that runs a simulation test with the given parameters.
+// It iterates over the provided seeds and runs the simulation test for each seed in parallel.
+//
+// It sets up the environment, creates an instance of the simulation app,
+// calls the simulation.SimulateFromSeed function to run the simulation, and performs post-run actions for each seed.
+// The execution is deterministic and can be used for fuzz tests as well.
+//
+// The system under test is isolated for each run but unlike the old runsim command, there is no Process separation.
+// This means, global caches may be reused for example. This implementation build upon the vanialla Go stdlib test framework.
+func RunWithSeeds[T SimulationApp](
```

Comment:
Why does this code need to be generic as opposed to just accepting the `SimulationApp` interface as arguments?

Reply:
With the generic used, the concrete type of the app can be passed to the `setupStateFactory` which accesses some fields and methods that would not be available via interface. It also enables the `NewSimApp` to be passed as parameter here without any changes. 
Alternatively, we can decorate the `NewSimApp` and do a cast in `setupStateFactory`. My preferences are with generics to avoid this.

Reply:
yeah this makes sense to me.  I just wanted to see if there was a simpler way we could use this